---
title: "CSHL Workshop Tutorial"
author: "Almas K."
date: "30/01/2021"
output:
  html_document:
    keep_md: true
    toc: true
    toc_depth: 3
    toc_float:
      collapsed: true
      smooth_scroll: true
    number_sections:  true
    theme: paper
    self_contained: yes
editor_options: 
  chunk_output_type: inline
---

## Accreditation 

Adapted from the following resources:

- Data Wrangling workshop for the BC Children's Trainee Omics Group (TOG) by Almas Khan 
- Also Adapted from Tutorials of Victor Yuan for the BC Children's TOG.
- 



## Setup

- Please have the latest version of R and Rstudio installed. 
- Download the [r markdown file]() for this workshop and open it in Rstudio.

- Install the following R packages now, if you haven't already:

```{r, eval = F}
install.packages(c("Bioconductor","swirl"))
### Add swirl Installation
```

## Overall Learning outcomes

- Knowledge of R and Bioconductor

- Knowledge of basic R data types

- Differentiate a matrix and data.frame 
- Understanding how to :
  - Assign variables
  - Run built in operations in R 
  - Subset on a data.frame and matrix

<!---The following chunk allows errors when knitting--->

```{r allow errors, echo = FALSE}
knitr::opts_chunk$set(error = TRUE)
```

## Additional Resources: 

- [R Swirl](https://swirlstats.com/) for interactive lesson on programming with R
- *R for data science* [tibbles chapter](http://r4ds.had.co.nz/tibbles.html) to learn more about tibble/data.frames
- [Learnr package](https://rstudio.github.io/learnr/) another interactive lesson within R on programming


I recommend using **r markdown file (recommended)** for all of your analysis scripts. Think of it as a like a notebook for your code that includes comments and I find it's easier to keep track of your analysis. 

In `.rmd` files, code needs to be in code chunks (insert code chunk shortcut is Windows: `ctrl` `+` `alt` `+` `i`, Mac: `cmd` `+` `alt` `+` `i`) in order to run.

```{r}

```

## Topics

**What is R?**

- What are vectors?
  - Basic data types
  
- `data.frame` and `matrix` objects
  - Basic functions for exploring dataframes
  - Basic functions for matrices
  - Indexing and Subsetting
- Common R operations
- Variables 

 **What is Bioconductor?**

- Introduction 

**What is BiocSwirl**

- Introduction
- Notes and topics 

### Data Types

In R there are multiple types of data and I have listed a couple of basic ones below: 

```{r}
3.14 # This is a numeric 
"This is a character" # A character type 
"3.14" # This is also a character 
TRUE # logical
FALSE # These are logicals
```

## Data frames and matrices


A data.frame is basically just a table, it has a certain number of rows, and a certain number of columns.

"trees" is a built-in dataset in R available as a `data.frame`.

```{r}
trees
```

The columns of the trees `data.frame` object are individual `vector` objects. So trees has 3 columns/vectors that are each 31 elements long. The dbl is just a type of numeric class of data. 

### Some basic functions to help understand your `data.frame` objects are:

```{r,error=TRUE}
# number of rows
nrow(trees)
# number of columns
ncol(trees)
# row x columns
dim(trees)
# some basic info on the "structure" of the data.frame
str(trees)
# calculates some summary statistics on each column
summary(trees)
# print first 6 rows
head(trees)
# print last 6 rows
tail(trees)
```

A matrix is ____

# Introduction to RNAseq data analysis

## Accreditation

The BiocSwirl interactive course was based on [this RNA-seq workflow](https://master.bioconductor.org/packages/release/workflows/vignettes/rnaseqGene/inst/doc/rnaseqGene.html) published by Love et al., updated in 2019.

## Overall Learning outcomes

- What is RNAseq?
- What kind of questions can be addressed using RNAseq?
- What is differential expression analysis?
- How to conduct differential expression analysis of RNAseq data in R

## Resources mentioned in the lecture
- A short overview over [gene expression](https://www.nature.com/scitable/topicpage/gene-expression-14121669/) published by Scitable (Nature)
- [RNA-Seq: a revolutionary tool for transcriptomics, Wang et al. 2009](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2949280/)
- An overview over [different RNAseq analysis workflows](https://www.cd-genomics.com/resourse-Bioinformatics-Workflow-of-RNA-Seq.html
)
- [The RNA-seq workflow used in this class for differential expression analysis](https://master.bioconductor.org/packages/release/workflows/vignettes/rnaseqGene/inst/doc/rnaseqGene.html) published by Love et al., updated in 2019.
- The [BiocSwirl](https://github.com/biocswirl-dev-team/BiocSwirl) documentation on github


## RNAseq workflow and course structure
![RNAseq workflow image](RNAseq.jpg)

## BiocSwirl course installation
```{r, eval = FALSE}
# installation 
install.packages("devtools")
library(devtools)
devtools::install_github("biocswirl-dev-team/BiocSwirl", build_vignettes = TRUE)
library(BiocSwirl)
library(swirl)
```

```{r, eval = FALSE}
# select RNAseq course
list_courses()
load_course('RNAseq')
start_course()
```

These following commands can be used within the BiocSwirl course environment.

```{r, eval = FALSE}
bye() #exits the course environment
play() #suspends feedback, allows you to ‘play’ with code
nxt() #continue feedback
skip() # skips question, not recommended
```



